File: ghrah-observer-webui/packages/protocol/src/payloads/agent.ts
import { z } from "zod";

export const AgentConfigPayloadSchema = z.object({
  name: z.string(),
  agent_id: z.string().optional(),
  agent_config_name: z.string().nullable().optional(),
  description: z.string().optional().default(""),
  system_prompt: z.string().optional().default(""),
  max_iterations: z.number().int().optional().default(10),
  communication_timeout: z.number().optional(),
  window: z.record(z.unknown()).nullable().optional(),
  context: z.record(z.unknown()).nullable().optional(),
  model_overrides: z.record(z.unknown()).nullable().optional(),
});

export const AbilityDefinitionPayloadSchema = z.object({
  ability_type: z.string(),
  params: z.record(z.unknown()).optional().default({}),
});

export const SpawnAgentPayloadSchema = z.object({
  project_id: z.string(),
  cluster_id: z.string().optional().default(""),
  config: AgentConfigPayloadSchema,
  abilities: z.array(AbilityDefinitionPayloadSchema).nullable().optional(),
  manifest_ref: z.string().nullable().optional(),
});

export const TerminateAgentPayloadSchema = z.object({
  project_id: z.string(),
  agent_id: z.string(),
  name: z.string(),
});

export const SendMessagePayloadSchema = z.object({
  project_id: z.string(),
  agent_id: z.string(),
  target: z.string(),
  content: z.string(),
  sender: z.string().optional().default("user"),
  timeout: z.number().nullable().optional(),
  metadata: z.record(z.unknown()).nullable().optional(),
});

export const BroadcastMessagePayloadSchema = z.object({
  project_id: z.string(),
  content: z.string(),
  sender: z.string().optional().default("user"),
});

export const RegisterAbilityPayloadSchema = z.object({
  project_id: z.string(),
  agent_id: z.string(),
  agent_name: z.string(),
  ability: AbilityDefinitionPayloadSchema,
});

export const UnregisterAbilityPayloadSchema = z.object({
  project_id: z.string(),
  agent_id: z.string(),
  agent_name: z.string(),
  ability_name: z.string(),
});

export const ListAgentsPayloadSchema = z.object({
  project_id: z.string(),
});

export const HealthCheckPayloadSchema = z.object({});

export const DelegatePayloadSchema = z.object({
  project_id: z.string(),
  from_agent_id: z.string(),
  to_agent_id: z.string(),
  from_agent: z.string(),
  to_agent: z.string(),
  content: z.string(),
  timeout: z.number().nullable().optional(),
});

export const GetAgentInfoPayloadSchema = z.object({
  project_id: z.string(),
  agent_id: z.string(),
  name: z.string(),
});

export const AgentCompactContextPayloadSchema = z.object({
  project_id: z.string(),
  agent_id: z.string(),
  agent_name: z.string(),
  cluster_id: z.string().optional().default(""),
});

export const SubscribePayloadSchema = z.object({
  agent_names: z.array(z.string()).nullable().optional(),
  event_types: z.array(z.string()).nullable().optional(),
});

export const UnsubscribePayloadSchema = z.object({
  agent_names: z.array(z.string()).nullable().optional(),
  event_types: z.array(z.string()).nullable().optional(),
});

export const ExecuteAbilityPayloadSchema = z.object({
  request_id: z.string(),
  agent_name: z.string(),
  ability_name: z.string(),
  tool_args: z.record(z.unknown()).optional().default({}),
});

export const HITLResponsePayloadSchema = z.object({
  promise_id: z.string().optional().default(""),
  agent_name: z.string().optional().default(""),
  ability_name: z.string().optional().default(""),
  tool_call_id: z.string().optional().default(""),
  approved: z.boolean(),
  reason: z.string().nullable().optional(),
  result: z.unknown().nullable().optional(),
});

export const InitClusterPayloadSchema = z.object({
  cluster_id: z.string(),
  config: z.record(z.unknown()).optional().default({}),
});

export const ShutdownClusterPayloadSchema = z.object({
  cluster_id: z.string(),
  config: z.record(z.unknown()).optional().default({}),
});

export const ClusterStatusPayloadSchema = z.object({
  cluster_id: z.string(),
});

export const ListClustersPayloadSchema = z.object({});

export const ClusterInfoPayloadSchema = z.object({
  cluster_id: z.string(),
  active_agents: z.number(),
  status: z.string(),
  bound: z.boolean(),
});

export const ListClustersResultPayloadSchema = z.object({
  clusters: z.array(ClusterInfoPayloadSchema),
});

export const AgentSpawnedPayloadSchema = z.object({
  project_id: z.string().optional().default(""),
  cluster_id: z.string().optional().default(""),
  name: z.string(),
  agent_id: z.string().optional().default(""),
  incarnation_id: z.string().optional().default(""),
  recovery_mode: z.string().optional().default(""),
  config: AgentConfigPayloadSchema,
});

export const AgentTerminatedPayloadSchema = z.object({
  project_id: z.string().optional().default(""),
  cluster_id: z.string().optional().default(""),
  name: z.string(),
  agent_id: z.string().optional().default(""),
  incarnation_id: z.string().optional().default(""),
});

export const ContentBlockSchema = z.discriminatedUnion("type", [
  z.object({
    type: z.literal("text"),
    text: z.string(),
  }),
  z.object({
    type: z.literal("reasoning"),
    reasoning: z.string(),
    incomplete: z.boolean().optional().default(false),
  }),
  z.object({
    type: z.literal("image"),
    url: z.string().nullable().optional(),
    base64: z.string().nullable().optional(),
    mime_type: z.string().nullable().optional(),
  }),
  z.object({
    type: z.literal("audio"),
    data: z.string(),
    mime_type: z.string(),
  }),
  z.object({
    type: z.literal("file"),
    url: z.string().nullable().optional(),
    base64: z.string().nullable().optional(),
    mime_type: z.string().nullable().optional(),
    filename: z.string().nullable().optional(),
  }),
  z.object({
    type: z.literal("tool_call"),
    id: z.string(),
    name: z.string(),
    arguments: z.record(z.unknown()),
  }),
  z.object({
    type: z.literal("tool_result"),
    tool_call_id: z.string(),
    name: z.string().nullable().optional(),
    content: z.string(),
    success: z.boolean().optional().default(true),
    error: z.string().nullable().optional(),
  }),
  z.object({
    type: z.literal("error"),
    error_type: z.string(),
    message: z.string(),
    details: z.record(z.unknown()).nullable().optional(),
  }),
]);

export type ContentBlock = z.infer<typeof ContentBlockSchema>;

export const ChatMessageWireSchema = z.object({
  role: z.enum(["system", "user", "ai", "tool"]).default("user"),
  content_blocks: z.array(ContentBlockSchema).optional().default([]),
  source: z.string().nullable().optional(),
  metadata: z.record(z.unknown()).optional().default({}),
});
export type ChatMessageWire = z.infer<typeof ChatMessageWireSchema>;

export const ActionResultItemSchema = z.object({
  ability_name: z.string(),
  action_result: z
    .object({
      outcome: z.enum(["success", "failure", "needs_input", "delegate"]).optional(),
      data: z.record(z.unknown()).optional().default({}),
      next_action_hint: z.string().nullable().optional(),
    })
    .nullable()
    .optional(),
});
export type ActionResultItem = z.infer<typeof ActionResultItemSchema>;

export const ActionNodeSchema = z.object({
  id: z.string().optional().default(""),
  parent_id: z.string().nullable().optional(),
  agent_name: z.string().optional().default(""),
  timestamp: z.string().optional(),
  iteration: z.number().int().optional().default(0),
  ability_names: z.array(z.string()).optional().default([]),
  agent_state: z.record(z.unknown()).optional().default({}),
  messages_delta: z.array(ChatMessageWireSchema).optional().default([]),
  messages_snapshot: z.array(ChatMessageWireSchema).nullable().optional(),
  is_snapshot: z.boolean().optional().default(false),
  action_results: z.array(ActionResultItemSchema).optional().default([]),
  metadata: z.record(z.unknown()).optional().default({}),
  session_id: z.string().optional().default(""),
  created_on_branch_id: z.string().optional().default(""),
});
export type ActionNode = z.infer<typeof ActionNodeSchema>;

export const AgentResponsePayloadSchema = z.object({
  project_id: z.string().optional().default(""),
  agent_id: z.string().optional().default(""),
  cluster_id: z.string().optional().default(""),
  sender: z.string(),
  recipient: z.string(),
  content: z.string(),
  content_blocks: z.array(ContentBlockSchema).nullable().optional(),
  message_type: z.string().optional().default("result"),
  metadata: z.record(z.unknown()).optional().default({}),
});

export const ActionChainUpdatedPayloadSchema = z.object({
  project_id: z.string().optional().default(""),
  agent_id: z.string().optional().default(""),
  cluster_id: z.string().optional().default(""),
  agent_name: z.string(),
  node: ActionNodeSchema.optional().default({}),
});

export const ContextUsageUpdatedPayloadSchema = z.object({
  project_id: z.string().optional().default(""),
  agent_id: z.string().optional().default(""),
  cluster_id: z.string().optional().default(""),
  agent_name: z.string(),
  phase: z.string(),
  occupied_tokens: z.number().int().nullable().optional(),
  basis: z.string(),
  budget_tokens: z.number().int().optional().default(0),
  budget_source: z.string().nullable().optional(),
  compact_threshold: z.number().nullable().optional(),
  real_input_tokens: z.number().int().nullable().optional(),
  real_output_tokens: z.number().int().nullable().optional(),
  real_cache_read_tokens: z.number().int().nullable().optional(),
  real_cache_write_tokens: z.number().int().nullable().optional(),
  compaction: z.record(z.unknown()).nullable().optional(),
  iteration: z.number().int().nullable().optional(),
});

export const AgentErrorPayloadSchema = z.object({
  project_id: z.string().optional().default(""),
  agent_id: z.string().optional().default(""),
  cluster_id: z.string().optional().default(""),
  agent_name: z.string(),
  error: z.string(),
});

export const HealthStatusPayloadSchema = z.object({
  status: z.record(z.boolean()),
});

export const AbilityResultPayloadSchema = z.object({
  project_id: z.string().optional().default(""),
  agent_id: z.string().optional().default(""),
  cluster_id: z.string().optional().default(""),
  request_id: z.string(),
  agent_name: z.string(),
  ability_name: z.string(),
  success: z.boolean(),
  result: z.unknown().nullable().optional(),
  error: z.string().nullable().optional(),
});

export const HITLRequestPayloadSchema = z.object({
  project_id: z.string().optional().default(""),
  agent_id: z.string().optional().default(""),
  cluster_id: z.string().optional().default(""),
  promise_id: z.string(),
  agent_name: z.string(),
  ability_name: z.string(),
  tool_args: z.record(z.unknown()).optional().default({}),
  context: z.record(z.unknown()).optional().default({}),
});

export type AgentConfigPayload = z.infer<typeof AgentConfigPayloadSchema>;
export type AbilityDefinitionPayload = z.infer<typeof AbilityDefinitionPayloadSchema>;
export type SpawnAgentPayload = z.infer<typeof SpawnAgentPayloadSchema>;
export type TerminateAgentPayload = z.infer<typeof TerminateAgentPayloadSchema>;
export type SendMessagePayload = z.infer<typeof SendMessagePayloadSchema>;
export type BroadcastMessagePayload = z.infer<typeof BroadcastMessagePayloadSchema>;
export type RegisterAbilityPayload = z.infer<typeof RegisterAbilityPayloadSchema>;
export type UnregisterAbilityPayload = z.infer<typeof UnregisterAbilityPayloadSchema>;
export type ListAgentsPayload = z.infer<typeof ListAgentsPayloadSchema>;
export type HealthCheckPayload = z.infer<typeof HealthCheckPayloadSchema>;
export type DelegatePayload = z.infer<typeof DelegatePayloadSchema>;
export type GetAgentInfoPayload = z.infer<typeof GetAgentInfoPayloadSchema>;
export type AgentCompactContextPayload = z.infer<typeof AgentCompactContextPayloadSchema>;
export type SubscribePayload = z.infer<typeof SubscribePayloadSchema>;
export type UnsubscribePayload = z.infer<typeof UnsubscribePayloadSchema>;
export type ExecuteAbilityPayload = z.infer<typeof ExecuteAbilityPayloadSchema>;
export type HITLResponsePayload = z.infer<typeof HITLResponsePayloadSchema>;
export type InitClusterPayload = z.infer<typeof InitClusterPayloadSchema>;
export type ShutdownClusterPayload = z.infer<typeof ShutdownClusterPayloadSchema>;
export type ClusterStatusPayload = z.infer<typeof ClusterStatusPayloadSchema>;
export type ListClustersPayload = z.infer<typeof ListClustersPayloadSchema>;
export type ClusterInfoPayload = z.infer<typeof ClusterInfoPayloadSchema>;
export type ListClustersResultPayload = z.infer<typeof ListClustersResultPayloadSchema>;

export type AgentSpawnedPayload = z.infer<typeof AgentSpawnedPayloadSchema>;
export type AgentTerminatedPayload = z.infer<typeof AgentTerminatedPayloadSchema>;
export type AgentResponsePayload = z.infer<typeof AgentResponsePayloadSchema>;
export type ActionChainUpdatedPayload = z.infer<typeof ActionChainUpdatedPayloadSchema>;
export type ContextUsageUpdatedPayload = z.infer<typeof ContextUsageUpdatedPayloadSchema>;
export type AgentErrorPayload = z.infer<typeof AgentErrorPayloadSchema>;
export type HealthStatusPayload = z.infer<typeof HealthStatusPayloadSchema>;
export type AbilityResultPayload = z.infer<typeof AbilityResultPayloadSchema>;
export type HITLRequestPayload = z.infer<typeof HITLRequestPayloadSchema>;

File: ghrah-protocol/src/ghrah/protocol/payloads/agent.py
# SPDX-FileCopyrightText: 2026 chenxya <chenxya@ghrah.org>
#
# SPDX-License-Identifier: Apache-2.0
"""Agent 域载荷模型。

Agent 生命周期/能力/订阅/HITL/集群命令与事件、上下文占用与 compact 命令载荷。
"""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field

# ─── 命令载荷模型 ───


class AgentConfigPayload(BaseModel):
    """Agent 配置载荷，对应 ghrah-core 的 AgentConfig。"""

    name: str
    agent_id: str = ""
    agent_config_name: str | None = None
    description: str = ""
    system_prompt: str = ""
    max_iterations: int = 10
    communication_timeout: float = 300.0
    window: dict[str, Any] | None = None
    context: dict[str, Any] | None = None
    model_overrides: dict[str, Any] | None = None


class SpawnAgentPayload(BaseModel):
    """spawn_agent 命令载荷。"""

    project_id: str
    cluster_id: str = ""
    config: AgentConfigPayload
    abilities: list[AbilityDefinitionPayload] | None = None
    manifest_ref: str | None = None


class TerminateAgentPayload(BaseModel):
    """terminate_agent 命令载荷。"""

    project_id: str
    agent_id: str
    name: str


class SendMessagePayload(BaseModel):
    """send_message 命令载荷。"""

    project_id: str
    agent_id: str
    target: str
    content: str
    sender: str = "user"
    timeout: float | None = None
    metadata: dict[str, Any] | None = None
    """扩展元数据（透传至 AgentMessage.metadata；如 Room 投递写入 room_id）。

    回复归属推导经链节点 messages_delta（receive 侧将 metadata 合入用户
    ChatMessage），非回复消息回传。
    """


class BroadcastMessagePayload(BaseModel):
    """broadcast_message 命令载荷。"""

    project_id: str
    content: str
    sender: str = "user"


class AbilityDefinitionPayload(BaseModel):
    """Ability 定义载荷。

    用于从 Subject 传输 Ability 配置到 Core。
    ability_type 映射到 ghrah-core 中内置的 Ability 类。
    """

    ability_type: str
    params: dict[str, Any] = Field(default_factory=dict)


class RegisterAbilityPayload(BaseModel):
    """register_ability 命令载荷。"""

    project_id: str
    agent_id: str
    agent_name: str
    ability: AbilityDefinitionPayload


class ListAgentsPayload(BaseModel):
    """list_agents 命令载荷（Project-scoped）。"""

    project_id: str


class HealthCheckPayload(BaseModel):
    """health_check 命令载荷（空载荷）。"""

    pass


class DelegatePayload(BaseModel):
    """delegate 命令载荷。"""

    project_id: str
    from_agent_id: str
    to_agent_id: str
    from_agent: str
    to_agent: str
    content: str
    timeout: float | None = None


class SubscribePayload(BaseModel):
    """subscribe 命令载荷。"""

    agent_names: list[str] | None = None  # None 表示订阅全部
    event_types: list[str] | None = None  # None 表示订阅全部事件类型


class UnsubscribePayload(BaseModel):
    """unsubscribe 命令载荷。"""

    agent_names: list[str] | None = None
    event_types: list[str] | None = None


class ExecuteAbilityPayload(BaseModel):
    """execute_ability 命令载荷。

    Core → Subject：请求 Subject 执行 Ability。
    """

    request_id: str
    agent_name: str
    ability_name: str
    tool_args: dict[str, Any] = Field(default_factory=dict)  # ─── 事件载荷模型 ───


class AgentSpawnedPayload(BaseModel):
    """agent_spawned 事件载荷。"""

    project_id: str = ""
    cluster_id: str = ""
    name: str
    agent_id: str = ""
    incarnation_id: str = ""
    recovery_mode: str = ""
    config: AgentConfigPayload


class AgentTerminatedPayload(BaseModel):
    """agent_terminated 事件载荷。"""

    project_id: str = ""
    cluster_id: str = ""
    name: str
    agent_id: str = ""
    incarnation_id: str = ""


class AgentResponsePayload(BaseModel):
    """agent_response 事件载荷。"""

    project_id: str = ""
    agent_id: str = ""
    cluster_id: str = ""
    sender: str
    recipient: str
    content: str
    content_blocks: list[dict[str, Any]] | None = None
    message_type: str = "result"
    metadata: dict[str, Any] = Field(default_factory=dict)


class ActionChainUpdatedPayload(BaseModel):
    """action_chain_updated 事件载荷。"""

    project_id: str = ""
    agent_id: str = ""
    cluster_id: str = ""
    agent_name: str
    node: dict[str, Any] = Field(default_factory=dict)


class AgentErrorPayload(BaseModel):
    """agent_error 事件载荷。"""

    project_id: str = ""
    agent_id: str = ""
    cluster_id: str = ""
    agent_name: str
    error: str


class HealthStatusPayload(BaseModel):
    """health_status 事件载荷。"""

    status: dict[str, bool]


class AbilityResultPayload(BaseModel):
    """ability_result 事件载荷。

    Core → Subject → Observer：Core 已确认收到的 Ability 执行结果事件。
    执行结果来源于 Subject 返回的 execute_ability command_result。
    """

    project_id: str = ""
    agent_id: str = ""
    cluster_id: str = ""
    request_id: str
    agent_name: str
    ability_name: str
    success: bool
    result: Any = None
    error: str | None = None


class HITLRequestPayload(BaseModel):
    """hitl_request 事件载荷。

    Subject → Observer：HITL 审批请求。
    当 AbilityRunner 判断某操作需要人工审批时，创建 Promise 并广播此事件。
    """

    project_id: str = ""
    agent_id: str = ""
    cluster_id: str = ""
    promise_id: str
    agent_name: str
    ability_name: str
    tool_args: dict[str, Any] = Field(default_factory=dict)
    context: dict[str, Any] = Field(default_factory=dict)


class ContextUsageUpdatedPayload(BaseModel):
    """context_usage_updated 事件载荷。

    Core → Subject → Observer：上下文占用状态感知出口（pre_call/post_call
    双相位）。occupied_tokens 为真实 usage 权威口径：pre_call 取占用锚点
    （可能为 None，如首次调用/head 变更后），post_call 取本次 LLM 调用
    真实 input_tokens（basis="real"，归一化口径：恒为总输入，含缓存
    部分）。budget_tokens 为落定预算（无 WindowManager 时为 0），来源由
    budget_source 标注：显式声明（declared）/ 模型窗口表（model_table）/
    内置默认（default）/ 厂商超限错误回填（vendor）。
    """

    project_id: str = ""
    agent_id: str = ""
    cluster_id: str = ""
    agent_name: str
    phase: str
    """发布相位："pre_call"（发送前，占用锚点口径）/ "post_call"（响应后，真实值口径）"""
    occupied_tokens: int | None = None
    basis: str
    """occupied_tokens 计量口径："anchor"（占用锚点）/"real"（本次调用真实值）"""
    budget_tokens: int = 0
    budget_source: str | None = None
    """预算来源："declared"/"model_table"/"default"/"vendor"（无 WindowManager 时为 None）"""
    compact_threshold: float | None = None
    real_input_tokens: int | None = None
    """本次调用总输入 token 数（归一化口径，含缓存命中与缓存写入部分）"""
    real_output_tokens: int | None = None
    real_cache_read_tokens: int | None = None
    """本次调用缓存命中读 token 数（pre_call 或厂商未上报时为 None/0）"""
    real_cache_write_tokens: int | None = None
    """本次调用缓存写入 token 数（pre_call 或厂商未上报时为 None/0）"""
    compaction: dict[str, Any] | None = None
    """本轮压缩决策记录（compaction_decision；未配置窗口时为 None）"""
    iteration: int | None = None


class HITLResponsePayload(BaseModel):
    """hitl_response 命令载荷。

    Observer → Subject（分布式）/ Observer → Core（单体）：HITL 审批响应。

    存在两条数据流，字段集合不同，故全部字段（除 approved）设默认值以兼容：
    - 分布式（Subject 侧）：promise_id / approved / reason
      （SubjectService._handle_hitl_response 据此 resolve Promise）
    - 单体（Core 侧）：agent_name / ability_name / tool_call_id / approved / result
      （core/router._handle_hitl_response 据此 resolve HITLFutureStore）
    """

    # 分布式路径字段
    promise_id: str = ""
    # 单体路径字段
    agent_name: str = ""
    ability_name: str = ""
    tool_call_id: str = ""
    # 两路径共有
    approved: bool
    # 分布式：拒绝原因；单体：审批附加结果
    reason: str | None = None
    result: Any = None


class UnregisterAbilityPayload(BaseModel):
    """unregister_ability 命令载荷。"""

    project_id: str
    agent_id: str
    agent_name: str
    ability_name: str


class GetAgentInfoPayload(BaseModel):
    """get_agent_info 命令载荷。"""

    project_id: str
    agent_id: str
    name: str


class AgentCompactContextPayload(BaseModel):
    """agent_compact_context 命令载荷。

    Observer → Subject → Core：手动触发链上 compact 回合。
    迭代中：置标志、下一节点提交前消费、进 compact 回合；空闲：直接执行
    一轮 compact；连续多次请求合并（幂等）。
    """

    project_id: str
    agent_id: str
    agent_name: str
    cluster_id: str = ""


class InitClusterPayload(BaseModel):
    """init_cluster 命令载荷。"""

    cluster_id: str
    config: dict[str, Any] = Field(default_factory=dict)


class ShutdownClusterPayload(BaseModel):
    """shutdown_cluster 命令载荷。"""

    cluster_id: str
    config: dict[str, Any] = Field(default_factory=dict)


class ClusterStatusPayload(BaseModel):
    """cluster_status 命令载荷。"""

    cluster_id: str


class ListClustersPayload(BaseModel):
    """list_clusters 命令载荷（空载荷）。"""

    pass


class ClusterInfoPayload(BaseModel):
    """单个集群的信息（list_clusters 结果项）。"""

    cluster_id: str
    active_agents: int
    status: str
    bound: bool


class ListClustersResultPayload(BaseModel):
    """list_clusters 命令结果载荷。"""

    clusters: list[ClusterInfoPayload]
